src/client/api/kernelApi.ts

```diff
@@ -96,26 +96,28 @@ class JupyterKernelService implements IExportedKernelService {
             disposables
         );
     }
-    async getKernelSpecifications(): Promise<KernelConnectionMetadata[]> {
+    async getKernelSpecifications(refresh?: boolean): Promise<KernelConnectionMetadata[]> {
         sendTelemetryEvent(Telemetry.JupyterKernelApiUsage, undefined, {
             extensionId: this.callingExtensionId,
             pemUsed: 'getKernelSpecifications'
         });
-        await this.notebookControllerManager.loadNotebookControllers();
+        await this.notebookControllerManager.loadNotebookControllers(refresh);
         const items = await this.notebookControllerManager.kernelConnections;
-        return items.map((item) => this.translateKernelConnectionMetataToExportedType(item));
+        return items.map((item) => this.translateKernelConnectionMetadataToExportedType(item));
     }
     async getActiveKernels(): Promise<{ metadata: KernelConnectionMetadata; notebook: NotebookDocument }[]> {
         sendTelemetryEvent(Telemetry.JupyterKernelApiUsage, undefined, {
             extensionId: this.callingExtensionId,
             pemUsed: 'getActiveKernels'
         });
-        return this.kernelProvider.kernels.map((item) => {
-            return {
-                metadata: this.translateKernelConnectionMetataToExportedType(item.kernelConnectionMetadata),
-                notebook: item.notebookDocument
-            };
-        });
+        return this.kernelProvider.kernels
+            .filter((item) => item.startedAtLeastOnce)
```

Comment:
See comments in the interface

Reply:
Here we should not return kernel objects that have not started
E.g. when opening a notebook we create a kernel and assign it to the notebook, however the kernel is not active, hence no point returning such kernels to the 3rd party extensions.